### PR TITLE
[To Feature] Back end changes for in-app role management

### PR DESCRIPTION
**Pull request by VineetBala-AOT**
Issue #: https://apps.itsm.gov.bc.ca/jira/browse/DESENG-574

*Description of changes:*
- Introduced new database models to manage user roles and user-to-role mappings in the Met database.
- Revamped the authentication validation process to rely on roles retrieved directly from the database, rather than relying solely on token-based roles.
- Eliminated redundant or unused roles from the system.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the met-public license (Apache 2.0).

### CODE REVIEW

File: met-api/src/met_api/utils/constants.py [hunk 1]
```diff
@@ -19,10 +19,10 @@
 class CompositeRoles(Enum):
     """Enumeration representing user roles."""
 
-    IT_ADMIN = 'Administrator'
+    ADMIN = 'Administrator'
```
Comment (Baelx) [inline met-api/src/met_api/utils/constants.py:22]: Good idea to rename these to remove the "IT"!

**Review by Baelx — APPROVED** [review]
Things look good!

I'm just curious: in terms of DB look-ups for checking user authentication. How many look-ups are we doing to very user roles? I'm curious about the long-term performance impacts

**Review by NatSquared — CHANGES_REQUESTED** [review]
Looks good... There is a problem with how you're accessing nested roles, though. Please have a look when you get a chance

File: met-api/src/met_api/__init__.py [hunk 2]
```diff
@@ -129,19 +132,40 @@ def setup_jwt_manager(app_context, jwt_manager):
...
-                                           'Please check your JWT_ROLE_CALLBACK '
-                                           'configuration.')
-                return []
-        return token_info
+            role = token_info.get(key, None)
+            if role is not None:  # Check if role is found
+                roles_from_token.extend(role)
+
+        # Check if 'create_tenant' role exists in token_info
+        if Role.CREATE_TENANT.value in roles_from_token:
```
[the remark this answers is not in the document]
Comment (NatSquared) [reply met-api/src/met_api/__init__.py:148]: To me it looks like this code is just using `CREATE_TENANT` as a standin superuser role until the actual superuser role is added to keycloak.

File: met-api/src/met_api/services/user_group_membership_service.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+from met_api.models.user_group import UserGroup
+from met_api.models.user_group_membership import UserGroupMembership
+from met_api.models.user_role import UserRole
+
+
+class UserGroupMembershipService:
+    """User group membership management service."""
+
+    @classmethod
+    def get_user_roles_within_a_tenant(cls, external_id, tenant_id) -> List[str]:
```
[the remark this answers is not in the document]
Comment (NatSquared) [reply met-api/src/met_api/services/user_group_membership_service.py:14]: Also, this function doesn't return a list of strings.
```suggestion
    def get_user_roles_within_tenant(cls, external_id, tenant_id) -> Tuple[List[str], int]:
```

File: met-api/src/met_api/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (VineetBala-AOT) [reply met-api/src/met_api/__init__.py:148]: Yes I will use the composite role once the superuser composite role is added. We will not use the granular role.

File: met-api/src/met_api/utils/roles.py [hunk 4]
```diff
@@ -23,12 +23,10 @@ class Role(Enum):
 
     # STAFF Based roles
     CREATE_TENANT = 'create_tenant'
-    VIEW_TENANT = 'view_tenant'
```
[the remark this answers is not in the document]
Comment (VineetBala-AOT) [reply met-api/src/met_api/utils/roles.py:26]: These roles we are not using anymore so I removed them.

File: met-api/src/met_api/utils/enums.py [hunk 5]
```diff
@@ -59,22 +59,22 @@ class LoginSource(Enum):
     IDIR = 'idir'
 
 
-class KeycloakPermissionLevels(Enum):
-    """Keycloak permission levels."""
+class CompositeRoleNames(Enum):
+    """Keycloak composite role names."""
```
[the remark this answers is not in the document]
Comment (VineetBala-AOT) [reply met-api/src/met_api/utils/enums.py]: Thanks, updated it as suggested.

File: met-api/src/met_api/utils/enums.py [hunk 6]
```diff
@@ -59,22 +59,22 @@ class LoginSource(Enum):
...
-    REVIEWER = 'Reviewer'
+    ADMIN = 'ADMIN'
+    VIEWER = 'VIEWER'
+    TEAM_MEMBER = 'TEAM_MEMBER'
+    REVIEWER = 'REVIEWER'
 
 
-class KeycloakCompositeRoleNames(Enum):
+class CompositeRoleId(IntEnum):
     """Keycloak composite role names."""
```
[the remark this answers is not in the document]
Comment (VineetBala-AOT) [reply met-api/src/met_api/utils/enums.py]: Thanks, updated it as suggested.

File: met-api/src/met_api/models/user_group_membership.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    group_id = db.Column(db.Integer, db.ForeignKey('user_group.id'), nullable=False)
+    tenant_id = db.Column(db.Integer, db.ForeignKey('tenant.id'), primary_key=True, nullable=False)
+    is_active = db.Column(db.Boolean, nullable=False)
+
+    __table_args__ = (
+        PrimaryKeyConstraint('id', 'staff_user_external_id', 'tenant_id'),
+    )
+
+    @classmethod
+    def get_group_by_user_id(cls, external_id, tenant_id):
```
[the remark this answers is not in the document]
Comment (VineetBala-AOT) [reply met-api/src/met_api/models/user_group_membership.py:31]: Thanks for the feedback! I've added the backref attribute to both user_group_membership and user_group, and also adjusted the user_group_membership_service code accordingly to accommodate these changes.

File: met-api/src/met_api/models/user_group_membership.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    @classmethod
+    def remove_user_from_group(cls, external_id, tenant_id, membership_data: dict) -> UserGroupMembership:
+        """Remove a user from group."""
+        membership: UserGroupMembership = UserGroupMembership.query.filter_by(
+            staff_user_external_id=external_id, tenant_id=tenant_id).first()
+        if membership:
+            # Update membership fields
+            for key in ['is_active']:
+                if key in membership_data:
+                    setattr(membership, key, False)
```
[the remark this answers is not in the document]
Comment (VineetBala-AOT) [reply met-api/src/met_api/models/user_group_membership.py]: Thanks, updated it as suggested.

File: met-api/src/met_api/services/user_group_membership_service.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (VineetBala-AOT) [reply met-api/src/met_api/services/user_group_membership_service.py]: Thanks, updated it as suggested.

File: met-api/src/met_api/__init__.py [hunk 9]
```diff
@@ -129,19 +132,40 @@ def setup_jwt_manager(app_context, jwt_manager):
...
+            # Append 'create_tenant' role to user_roles list
+            user_roles = [Role.CREATE_TENANT.value]
+        else:
+            user_roles = []
```
[the remark this answers is not in the document]
Comment (VineetBala-AOT) [reply met-api/src/met_api/__init__.py]: Thanks, updated it as suggested.

File: met-api/src/met_api/__init__.py [hunk 10]
```diff
@@ -129,19 +132,40 @@ def setup_jwt_manager(app_context, jwt_manager):
...
-            token_info = token_info.get(key, None)
-            if token_info is None:
-                app_context.logger.warning('Unable to find role in token_info. '
-                                           'Please check your JWT_ROLE_CALLBACK '
-                                           'configuration.')
-                return []
-        return token_info
+            role = token_info.get(key, None)
+            if role is not None:  # Check if role is found
+                roles_from_token.extend(role)
```
[the remark this answers is not in the document]
Comment (VineetBala-AOT) [reply met-api/src/met_api/__init__.py]: Thanks, updated it as suggested.

**Review by NatSquared — APPROVED** [review]
Thanks for addressing those comments!

**Pull request merged**